### ISSUE TITLE
Suggest using `String::extend` where appropriate

### ISSUE THREAD

**Issue by JarredAllen** [rust-lang/rust-clippy#17633]
### What it does

The `String` type has some `Extend` implementations, and the lint suggests using those instead of pushing in a loop.

### Advantage

- More concise
- I suspect it optimizes better

### Example

```rust
for c in s.chars() {
    out_string.push(c);
}
```

Could be written as:

```rust
out_string.extend(s.chars());
```

And similar for e.g. `String::push_str` in a loop

### Comparison with existing lints

The closest related lint I can find is `string_extend_chars`, which would replace the output for the example I provided with `out_string.push_str(s);`. For other `Iterator<char>`s (e.g. [`core::char::ToLowercase`](https://doc.rust-lang.org/core/char/struct.ToLowercase.html) was the iterator I was using when inspired to make this issue ticket), the `String::extend` call would most likely be the last suggestion we can make.

### Additional Context

It might be reasonable to extend to other data structures beyond just `String`s.

**Comment by CommanderStorm** [comment]
Adding this false negative to `clippy::string_extend_chars` seems like a sensible option, or? This seems like essentially the same case, maybe with a bit of simple controll flow logic pattern matching mixed in..

**Comment by JarredAllen** [comment]
No, that lint specifically applies to the [`std::str::Chars`](https://doc.rust-lang.org/stable/std/str/struct.Chars.html) iterator, hence the name. This lint can apply regardless of type of iterator, so long as its item is a type that `String` implements extend for (which can be `char`, `&str`, `String`, or others).

The code I was looking at that made me suggest this was something like:

```rust
let mut s: String = ..;
let c: char = ..;
if foo(c) {
    for ch in c.to_lowercase() {
        s.push(ch);
    }
}
```

With the present standard library, there is no way to translate this to something that uses `String::push_str`, which [`string_extend_chars`](https://rust-lang.github.io/rust-clippy/master/index.html?search=extend_#string_extend_chars) recommends. The `ToLowercase` iterator type provides no way to get the remaining characters as a `&str` (and, by my read of the stdlib, it has a `[char; 3]` internally, not an` &str`, so it can't give that out even if we wanted that). And even if the stdlib made a way to get that iterator's output as a `&str`, there's still a functionally infinite amount of `char` or `&str` iterators, many of which can't be converted.

**Comment by blyxyas** [comment]
Regardless of readability, the extend method produces more instructions that the loop.

[See Compiler Explorer](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:rust,selection:(endColumn:38,endLineNumber:17,positionColumn:38,positionLineNumber:17,selectionStartColumn:38,selectionStartLineNumber:17,startColumn:38,startLineNumber:17),source:'//+Type+your+code+here,+or+load+an+example.%0A%0A//+As+of+Rust+1.75,+small+functions+are+automatically%0A//+marked+as+%60%23%5Binline%5D%60+so+they+will+not+show+up+in%0A//+the+output+when+compiling+with+optimisations.+Use%0A//+%60%23%5Bunsafe(no_mangle)%5D%60+or+%60%23%5Binline(never)%5D%60+to%0A//+work+around+this+issue.%0A//+See+https://github.com/compiler-explorer/compiler-explorer/issues/5939%0A%23%5Bunsafe(no_mangle)%5D%0Apub+fn+push_in_loop(first_string:+%26mut+String,+to_add:+String)+%7B%0A++++for+c+in+to_add.chars()+%7B%0A++++++++first_string.push(c)%3B%0A++++%7D%0A%7D%0A%0A%23%5Bunsafe(no_mangle)%5D%0Apub+fn+push_extend(first_string:+%26mut+String,+to_add:+String)+%7B%0A++++first_string.extend(to_add.chars())%3B%0A%7D%0A%0A%0A//+If+you+use+%60main()%60,+declare+it+as+%60pub%60+to+see+it+in+the+output:%0A//+pub+fn+main()+%7B+...+%7D%0A'),l:'5',n:'0',o:'Rust+source+%231',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:r1980,filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1',verboseDemangling:'0'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:rust,libs:!(),options:'-Copt-level%3D3',overrides:!((name:edition,value:'2024')),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+rustc+1.98.0+(Editor+%231)',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)

This might still be worth it, just not because of optimization.
